DOCSUP-3570: Update the SQL-statement SELECT (#16318)
ClickHouse/ClickHouse: docs/en/sql-reference/statements/select/order-by.md

│ 1970-03-12 │ 1970-01-08 │ original │
└────────────┴────────────┴──────────┘
```

## OFFSET FETCH Clause {#offset-fetch}

> `OFFSET` and `FETCH` allow you to retrieve just a portion of the rows that are generated by the rest of the query.

Reviewer: 1. There is no "rest of the query" after this clause. OFFSET FETCH is a final part of a query that limits the number of rows.
2. What is the motivation or a usecase for a user to use this clause?

``` sql

> OFFSET offset_row_count {ROW | ROWS} FETCH {FIRST | NEXT} fetch_row_count {ROW | ROWS} {ONLY | WITH TIES}

Reviewer: Does FETCH the mandatory part of a query? Can we use OFFSET separatedly? If FETCH is an optional part of a query you must show it in the syntax description by square brackets.

```

> The `FETCH` is an alternative to the [LIMIT] (../../../sql-reference/statements/select/limit.md) clause and retrieves rows from a query `SELECT` type.

Reviewer: 1. Link to LIMIT dosn't work.
2. FETCH is not a query, it doesn't retrieve rows, it specifies the maximum number of rows that can be in the result of a query.
3. You are describing the SELECT...ORDER BY query. No need to emphasize the type.

> The `OFFSET` says to skip that many rows before beginning to return rows.

Reviewer: The man says, the OFFSET specifies.

For example, the following query

``` sql

> SELECT * FROM test_fetch ORDER BY a OFFSET 1 ROW FETCH FIRST 3 ROWS ONLY

Reviewer: Queries must end with a semicolon `;`. Here and everywhere in examples.

```

is identical to the query

``` sql
SELECT * FROM test_fetch ORDER BY a LIMIT 3 OFFSET 1
```

> When using `FETCH`, it is important to use an [ORDER BY] (../../../sql-reference/statements/select/order-by.md) clause that constrains the result rows into a unique order. Otherwise, you will get an unpredictable subset of the query's rows.

Reviewer: Link to ORDER BY doesn't work.

`ROW` and `ROWS` as well as `FIRST` and `NEXT` are noise words that don't influence the effects of these conditions.

The `ONLY` option is used to return rows that immediately follow the rows omitted by the `OFFSET`.

> The `WITH TIES` option is used to return any additional rows that tie for the last place in the result set according to the `ORDER BY` clause. The `ORDER BY` is mandatory in this case.

Reviewer: ORDER BY is mandatory when you use it indeed. Does this sentece mean that OFFSET FETCH can be used as a separate clause and it doesn't neet ORDER BY at all? If OFFSET FETCH is independent from ORDER BY, why are you desribing it in the ORDER BY section?

!!! note "Note"
    According to the standard, the `OFFSET` clause must come before the `FETCH` clause if both are present.
	

> ### Example {#example}

Reviewer: It would be very good to show the difference between queries with ONLY and WITH TIES.

Input table:

``` text
┌─a─┬─b─┐
│ 1 │ 1 │
│ 2 │ 1 │
│ 3 │ 4 │
│ 1 │ 3 │
│ 5 │ 4 │
│ 0 │ 6 │
│ 5 │ 7 │
└───┴───┘
```

Query:

``` sql
SELECT * FROM test_fetch ORDER BY a OFFSET 1 ROW FETCH FIRST 3 ROWS ONLY 
```

Result:

``` text
┌─a─┬─b─┐

> │ 1 │ 1 │

Reviewer: Why do I see this string in the result. It must be skipped by OFFSET 1, doesn't it?

Author: > Why do I see this string in the result. It must be skipped by OFFSET 1, doesn't it?

No. After sorting, the first row is a = 0 and b = 6. This line is skipped.

│ 1970-03-12 │ 1970-01-08 │ original │
└────────────┴────────────┴──────────┘
```

## OFFSET FETCH Clause {#offset-fetch}

> `OFFSET` and `FETCH` are optional parts of the query, which allow you to retrieve just a portion of the rows.

Reviewer: ```suggestion
`OFFSET` and `FETCH` allow you to retrieve data by portions. They specify row block which you want to get by a single query.
```

``` sql

> [OFFSET offset_row_count {ROW | ROWS}] [FETCH {FIRST | NEXT} fetch_row_count {ROW | ROWS} {ONLY | WITH TIES}]

Reviewer: ```suggestion
OFFSET offset_row_count {ROW | ROWS}] [FETCH {FIRST | NEXT} fetch_row_count {ROW | ROWS} {ONLY | WITH TIES}]
```

```

> The `offset_row_count` or `fetch_row_count` value can be a literal constant, a parameter, or a variable name. If `fetch_row_count` is omitted, it defaults to 1.

Reviewer: ```suggestion
The `offset_row_count` or `fetch_row_count` value can be a number or a literal constant. You can omit  `fetch_row_count`; by default it equals 1.
```

> The `OFFSET` specifies to skip that many rows before beginning to return rows.

Reviewer: ```suggestion
`OFFSET` specifies the number of rows to skip before starting to return rows from the query.
```

The `FETCH` specifies the maximum number of rows that can be in the result of a query.

The `ONLY` option is used to return rows that immediately follow the rows omitted by the `OFFSET`. In this case the `FETCH` is an alternative to the [LIMIT](../../../sql-reference/statements/select/limit.md) clause. For example, the following query

``` sql
SELECT * FROM test_fetch ORDER BY a OFFSET 1 ROW FETCH FIRST 3 ROWS ONLY;
```

is identical to the query

``` sql
SELECT * FROM test_fetch ORDER BY a LIMIT 3 OFFSET 1;
```

But unlike `LIMIT` the `FETCH` has the `WITH TIES` option which used to return any additional rows that tie for the last place in the result set according to the `ORDER BY` clause.

> `ROW` and `ROWS` as well as `FIRST` and `NEXT` are noise words that don't influence the effects of these conditions.

Reviewer: Delete it, please.
